leshan-core/src/main/java/org/eclipse/leshan/core/request/argument/Arguments.java

```diff
@@ -0,0 +1,331 @@
+    }
+
+    /**
+     * Returns number of arguments.
+     */
+    public int size() {
+        return argumentMap.size();
+    }
+
+    /**
+     * Check if Arguments has any elements.
+     */
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    /**
+     * Checks if {@link Arguments} has {@link Argument} with digit.
+     */
+    public boolean hasDigit(int digit) {
+        validateDigit(digit);
+        return argumentMap.containsKey(digit);
+    }
+
+    /**
+     * Gets {@link Argument} by digit.
+     *
+     * @throws IllegalArgumentException if digit doesn't exists in {@link Arguments}
+     */
+    public Argument get(int digit) {
+        validateDigit(digit);
+        return argumentMap.get(digit);
+    }
+
+    /**
+     * Gets digit keys of arguments.
+     */
+    public Set<Integer> getDigits() {
+        return argumentMap.keySet();
+    }
+
+    /**
+     * Gets collection of {@link Argument}.
+     */
+    public Collection<Argument> getValues() {
+        return argumentMap.values();
+    }
+
+    @Override
+    public Iterator<Argument> iterator() {
+        return argumentMap.values().iterator();
+    }
+
+    private void validateDigit(int digit) throws IllegalArgumentException {
+        if (digit < 0 || digit > 9) {
+            throw new IllegalArgumentException(String.format("Invalid digit [%s]", digit));
+        }
+    }
+
+    /**
+     * Parse String into {@link Arguments}
+     *
+     * <pre>
+     * {@code
+     * arglist = arg *( "," arg )
+     * arg = DIGIT / DIGIT "=" "'" *CHAR "'"
+     * DIGIT = "0" / "1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9"
+     * CHAR = "!" / %x23-26 / %x28-5B / %x5D-7E
+     * }
+     * </pre>
+     * 
+     * @param arglist String contains arguments to parse.
+     * @return {@link Arguments} represents list of arguments.
+     * @throws InvalidArgumentException if text has invalid format.
+     */
+    public static Arguments parse(String arglist) throws InvalidArgumentException {
+        ArgumentsBuilder builder = builder();
+
+        if (arglist != null && !arglist.isEmpty()) {
+            int cursor = 0;
+            while (true) {
+                // consume argument
+                cursor = consumeArgument(arglist, cursor, builder);
+
+                // no more argument we finished
+                if (cursor == arglist.length()) {
+                    break;
+                }
+
+                // else consume separator ','
+                if (arglist.charAt(cursor) == ',') {
+                    cursor++;
+                } else {
+                    throw new InvalidArgumentException(
+                            "Unable to parse Arguments [%s] : [,] separator expected at index %d after [%s]", arglist,
+                            cursor, arglist.substring(0, cursor));
+                }
+            }
+        }
+        // Use builder capability to validate arguments uniqueness.
+        return builder.build();
```

Comment:
Here there is a little issue, when build failed we don't have the whole `arglist` in the exception message.
I guess it's not a big deal and I'm not sure there is an elegant way to do that.

Reply:
I think it's general problem with validation. If we use for validation other sub-validations we end up with more specific error message. We could catch and repackage exception to add more context. I also tried to replace throwing validation with test if object is correct - isValid() method which returns error code or error message.

Reply:
In this particular case, do you think we should "catch and repackage exception" to add more context ? or this is not worth it ?

Reply:
For me repackaging exceptions are not big deal - it simplifies code. From performance point of view - invalid data should be considered as an exception rather as regular usage.

Reply:
> From performance point of view - invalid data should be considered as an exception rather as regular usage.

I agree but I guess **in case where** there is really a big performance drawback which can be trigger by foreign peer. This could become an easy way to DoS. 

But in this particular case, I don't know and I even guess with the [default JVM behavior](https://www.oracle.com/java/technologies/javase/release-notes-introduction.html) about stacktrace this should not be an issue : 
```
The compiler in the server VM now provides correct stack backtraces for all "cold" built-in exceptions. For performance purposes, when such an exception is thrown a few times, the method may be recompiled. After recompilation, the compiler may choose a faster tactic using preallocated exceptions that do not provide a stack trace.
```

So I guess we can go this of rule: "if this is rather easy to avoid to repackaging we do it. If this is impossible or make the code really ugly in a design point of view, we repackage exception."

So in this case, I will repackage the builder exception in parse method to get a better exception message.